DataFormats/Common/interface/getThinned_implementation.h

```diff
@@ -0,0 +1,171 @@
+#ifndef DataFormats_Common_getThinned_implementation_h
+#define DataFormats_Common_getThinned_implementation_h
+
+#include <optional>
+#include <tuple>
+
+#include "DataFormats/Common/interface/ThinnedAssociation.h"
+#include "DataFormats/Provenance/interface/BranchID.h"
+#include "DataFormats/Provenance/interface/ProductID.h"
+#include "DataFormats/Provenance/interface/ThinnedAssociationsHelper.h"
+
+namespace edm {
+  class WrapperBase;
+
+  namespace detail {
+    // This function provides a common implementation of
+    // EDProductGetter::getThinnedProduct() for EventPrincipal,
+    // DataGetterHelper, and BareRootProductGetter.
+    //
+    // getThinnedProduct assumes getIt was already called and failed to find
+    // the product. The input key is the index of the desired element in the
+    // container identified by ProductID (which cannot be found).
+    // If the return value is not null, then the desired element was
+    // found in a thinned container. If the desired element is not
+    // found, then an optional without a value is returned.
+    template <typename F1, typename F2, typename F3>
+    std::optional<std::tuple<WrapperBase const*, unsigned int> > getThinnedProduct(
+        ProductID const& pid,
+        unsigned int key,
+        ThinnedAssociationsHelper const& thinnedAssociationsHelper,
+        F1 pidToBid,
+        F2 getThinnedAssociation,
+        F3 getByProductID) {
+      BranchID parent = pidToBid(pid);
+
+      // Loop over thinned containers which were made by selecting elements from the parent container
+      for (auto associatedBranches = thinnedAssociationsHelper.parentBegin(parent),
+                iEnd = thinnedAssociationsHelper.parentEnd(parent);
+           associatedBranches != iEnd;
+           ++associatedBranches) {
+        ThinnedAssociation const* thinnedAssociation = getThinnedAssociation(associatedBranches->association());
+        if (thinnedAssociation == nullptr)
+          continue;
+
+        if (associatedBranches->parent() != pidToBid(thinnedAssociation->parentCollectionID())) {
+          continue;
+        }
+
+        // Does this thinned container have the element referenced by key?
+        auto thinnedIndex = thinnedAssociation->getThinnedIndex(key);
+        if (not thinnedIndex.has_value()) {
```

Comment:
Wouldn't it be simpler to just do `if(not thinnedIndex)` since `std::optional` has an `operator bool`?

Reply:
It would be simpler. On the other hand, with `auto` declaration it would be less clear that `thinnedIndex` is `optional` (ok, I could change the `auto` to `std::optional<unsigned int>` or just `std::optional`).